### PR TITLE
ENH: Index.infer_objects

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -6521,6 +6521,31 @@ def drop(
...
+    def infer_objects(self, copy: bool = True) -> Index:
+        """
+        If we have an object dtype, try to infer a non-object dtype.
+
+        Parameters
+        ----------
+        copy : bool, default True
+            Whether to make a copy in cases where no inference occurs.
+        """
+        if self.dtype != object or self._is_multi:
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:6533]: Should MultiIndex just raise a `NotImplementedError` instead and just suggest to use `to_frame().infer_objects()`?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:6533]: good idea, will update.

**Review by mroeschke — APPROVED** [review]
LGTM. Can merge on green

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**